### ISSUE TITLE
Pyright fails to bind function arguments to type var tuple

### ISSUE THREAD

**Issue by LeeeeT** [microsoft/pyright#7820]
You've successfully defeated the [bug with the `curry` function](https://github.com/microsoft/pyright/issues/7588) and passed to the next level. On Level 2, your enemy is a sibling of `curry`, the `uncurry` function, which again causes a false positive.

```python
from collections.abc import Callable

def curry[First, *Rest, Result](function: Callable[[First, *Rest], Result]) -> Callable[[*Rest], Callable[[First], Result]]:
    return lambda *rest: lambda first: function(first, *rest)

def uncurry[First, *Rest, Result](function: Callable[[*Rest], Callable[[First], Result]]) -> Callable[[First, *Rest], Result]:
    return lambda first, *rest: function(*rest)(first)

@curry
@curry
def f(a: str, b: int, c: bool) -> None:
    pass

reveal_type(f)  # (bool) -> ((int) -> ((str) -> None))

uncurried = uncurry(uncurry(f))
```

```
file.py:16:21 - error: Argument of type "(int, *int) -> ((str) -> None)" cannot be assigned to parameter "function" of type "(*Rest@uncurry) -> ((First@uncurry) -> Result@uncurry)" in function "uncurry"
  Type "(int, *int) -> ((str) -> None)" is incompatible with type "(*Union[*Rest@uncurry], *Union[*Rest@uncurry]) -> ((str) -> None)"
file.py:16:29 - error: Argument of type "(bool) -> ((int) -> ((str) -> None))" cannot be assigned to parameter "function" of type "(*Rest(1)@uncurry) -> ((First(1)@uncurry) -> Result(1)@uncurry)" in function "uncurry"
  Type "(bool) -> ((int) -> ((str) -> None))" is incompatible with type "(*Union[*Rest@uncurry]) -> ((Union[*Rest@uncurry]) -> ((str) -> None))"
    Function return type "(int) -> ((str) -> None)" is incompatible with type "(Union[*Rest@uncurry]) -> ((str) -> None)"
      Type "(int) -> ((str) -> None)" is incompatible with type "(Union[*Rest@uncurry]) -> ((str) -> None)"
```

It, again, works without error if you apply `uncurry` once at a time like this:

```python
applied_once = uncurry(f)
applied_twice = uncurry(applied_once)

reveal_type(applied_once)  # (int, bool) -> ((str) -> None)
reveal_type(applied_twice)  # (str, int, bool) -> None
```

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.363, which I just published.
